lib/internal/fs/dir.js

```diff
@@ -0,0 +1,162 @@
+'use strict';
+
+const pathModule = require('path');
+const binding = internalBinding('fs');
+const dirBinding = internalBinding('fs_dir');
+const {
+  codes: {
+    ERR_MISSING_ARGS
+  }
+} = require('internal/errors');
+
+const { FSReqCallback } = binding;
+const internalUtil = require('internal/util');
+const {
+  getDirent,
+  getOptions,
+  getValidatedPath,
+  handleErrorFromBinding,
+  makeCallback
+} = require('internal/fs/utils');
+
+const kDirHandle = Symbol('Dir handle');
+const kDirPath = Symbol('Dir path');
+const kDirOptions = Symbol('Dir options');
+const kDirReadP = Symbol('Dir Awaitable Read');
+const kDirCloseP = Symbol('Dir Awaitable Close');
+
+class Dir {
+  constructor(handle, path, options) {
+    if (handle == null) throw new ERR_MISSING_ARGS('handle');
+    this[kDirHandle] = handle;
+    this[kDirPath] = path;
+
+    this[kDirOptions] = getOptions(options, {
+      encoding: 'utf8'
+    });
+
+    this[kDirReadP] = internalUtil.promisify(this.read.bind(this));
+    this[kDirCloseP] = internalUtil.promisify(this.close.bind(this));
+  }
+
+  get path() {
+    return pathModule.toNamespacedPath(this[kDirPath]);
+  }
+
+  read(options, callback) {
+    callback = typeof options === 'function' ? options : callback;
+    if (!callback) {
+      return this[kDirReadP](options);
+    }
+    callback = makeCallback(callback);
```

Comment:
As the comment for `makeCallback` says, we shouldn’t do this, because `callback` is not passed to the binding directly; `makeCallback` creates a closure each time it’s used, so it’s not free.